source/elements/oneMKL/source/domains/spblas/data_types/destroy_sparse_matrix.rst

```diff
@@ -0,0 +1,55 @@
+.. SPDX-FileCopyrightText: 2024 Intel Corporation
+..
+.. SPDX-License-Identifier: CC-BY-4.0
+
+.. _onemkl_sparse_destroy_sparse_matrix:
+
+destroy_sparse_matrix
+=====================
+
+Destroys a ``sparse_matrix_handle_t`` object.
+
+.. rubric:: Description and Assumptions
+
+The ``oneapi::mkl::sparse::destroy_sparse_matrix`` function frees the resources
+allocated for the handle.
+
+If a buffer was provided, its reference count is decremented.
+
+If a USM pointer was provided, the data is not free'd.
+
+.. rubric:: Syntax
+
+.. code-block:: cpp
+
+   namespace oneapi::mkl::sparse {
+
+       void destroy_sparse_matrix (sycl::queue                                 &queue,
+                                   oneapi::mkl::sparse::sparse_matrix_handle_t smhandle);
```

Comment:
A global comment on the synchronous functions. I have removed the `dependencies` and `sycl::event` from non-optimize and non-compute functions after all. My thinking is that these functions technically don't need a host_task to be implemented. With this API users have the choice to wrap this function inside a host_task, and potentially wait on SYCL events if needed. Some use-cases may not need asynchronous `destroy_*` so avoiding host_task will reduce the overhead.

Reply:
Key question for making a decision here is if one can have nested event submissions, e.g., can you have `queue.submit()` _inside_ a `host_task`; I am not aware that it works, and if it does, I'd like to see a code snippet of how that's done. Backend(s) such as Intel oneMKL can have event submissions in their `optimize_*` or `destroy_*` calls, so wrapping these APIs inside a `host_task` makes it a nested event submission. Wrapping these APIs in a `host_task` also means the `queue` may need to be (must be?) passed in by reference so that it can be passed into the backend as an argument.

Assuming nested event submission doesn't work, the workaround in the current proposed version of your API is to have a `queue.wait()` call immediately inside the function before dispatching to the backend, which may be an overkill as only some events may need to be waiting on. If we were to pass in `dependencies` but not return a `sycl::event`, that still forces the API to be synchronous, while allowing the backend to call `dependencies[i].wait()` rather than `queue.wait()`.

In essence, I believe that `dependencies` argument is necessary here and we cannot allow wrapping this into a `host_task`. To indicate that the API is synchronous for USM, we just shouldn't return a `sycl::event`, that's all.

I'm going to having to think a bit more about what happens with the `sycl::buffer` APIs.

Reply:
That is a very good point. The spec says that using a `queue` inside a `host_task` is undefined behavior. It makes sense for these functions to take event dependencies, wait on them and not return an event in that case. That way we can still avoid the `host_task`.
For the buffer API, to guarantee that the object won't be destroyed while buffers are in used we will have to submit an empty `host_task` that depends on the buffers and wait on it.
Also related to this [comment](https://github.com/oneapi-src/oneAPI-spec/pull/522#discussion_r1516581741), I think we could add a list of event dependencies for the buffer API.

Reply:
This is also a topic worth us discussing in person.

Reply:
I made `destroy_*` and `release_*` functions asynchronous in https://github.com/oneapi-src/oneAPI-spec/pull/522/commits/02013d5e70d92cb25284f64da5413132e1045dc5.
I think `create_*`, `init_*` and `set_*` can be kept synchronous.